### DISCUSSION TITLE
How to set favicon?

### DISCUSSION THREAD

**Discussion by Anarchist-byte** [Minestom/Minestom#835] [Help] [answered]
I don't understand how ResponseData#setFavicon works.
![image](https://user-images.githubusercontent.com/70373707/160876859-29342c33-5e5c-47ec-9452-61e5f2571db1.png)
I googled it a bit and found some online base64 decoders, so I did like "data:image/png;base64, <very-long-thing" in String of image, but it doesn't work.
My code looks something like this now:
![image](https://user-images.githubusercontent.com/70373707/160877049-ec9ea80c-cef5-4bb9-a3a5-cb3b787195be.png)
How can I set favicon?

**Comment by TheMode** [comment]
See https://wiki.vg/Server_List_Ping#Response
> The favicon should be a [PNG](http://en.wikipedia.org/wiki/Portable_Network_Graphics) image that is [Base64](http://en.wikipedia.org/wiki/Base64) encoded (without newlines: \n, new lines no longer work since 1.13) and prepended with data:image/png;base64,. It should also be noted that the source image must be exactly 64x64 pixels, otherwise the Notchian client will not render the image.

**Comment by emortaldev** [comment, marked as answer]
I would not recommend doing this manually, instead just put a png file on the server and read it

To convert a BufferedImage to a Base64 in Java, you can use this:
```java
BufferedImage image = ImageIO.read(new File("./logo.png"));
ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
ImageIO.write(image, "png", outputStream);
String base64String = Base64.getEncoder().encodeToString(outputStream.toByteArray());
outputStream.close();

// base64String contains what you want for your ResponseData favicon
```

**Reply by emortaldev** [reply]
This results in something like `iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAIAAAAlC+aJAAARIkl...`

**Reply by Anarchist-byte** [reply]
And next I use it like:
`responseData.setFavicon("data:image/png;base64, " + base64String);`

**Reply by Anarchist-byte** [reply]
Right?

**Reply by vanjmali** [reply]
Yes

**Reply by Anarchist-byte** [reply]
Thanks!

**Reply by dev-hydrogen** [reply]
> And next I use it like: `responseData.setFavicon("data:image/png;base64, " + base64String);`

It should not have a space after "base64,"
